rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
-{
-    pub fn len(&self) -> usize {
-        self.len
+pub type ArrayRef = Arc<Array>;
+
+/// Constructs an array using the input `data`. Returns a reference-counted `Array`
+/// instance.
+fn make_array(data: ArrayDataRef) -> ArrayRef {
+    // TODO: refactor the DataType enum and remove the clone here.
+    match data.data_type() {
+        DataType::Boolean => Arc::new(PrimitiveArray::<bool>::from(data)) as ArrayRef,
+        DataType::Int8 => Arc::new(PrimitiveArray::<i8>::from(data)) as ArrayRef,
+        DataType::Int16 => Arc::new(PrimitiveArray::<i16>::from(data)) as ArrayRef,
+        DataType::Int32 => Arc::new(PrimitiveArray::<i32>::from(data)) as ArrayRef,
+        DataType::Int64 => Arc::new(PrimitiveArray::<i64>::from(data)) as ArrayRef,
+        DataType::UInt8 => Arc::new(PrimitiveArray::<u8>::from(data)) as ArrayRef,
+        DataType::UInt16 => Arc::new(PrimitiveArray::<u16>::from(data)) as ArrayRef,
+        DataType::UInt32 => Arc::new(PrimitiveArray::<u32>::from(data)) as ArrayRef,
+        DataType::UInt64 => Arc::new(PrimitiveArray::<u64>::from(data)) as ArrayRef,
+        DataType::Float32 => Arc::new(PrimitiveArray::<f32>::from(data)) as ArrayRef,
+        DataType::Float64 => Arc::new(PrimitiveArray::<f64>::from(data)) as ArrayRef,
+        DataType::Utf8 => Arc::new(BinaryArray::from(data)) as ArrayRef,
+        DataType::List(_) => Arc::new(ListArray::from(data)) as ArrayRef,
+        DataType::Struct(_) => Arc::new(StructArray::from(data)) as ArrayRef,
+        dt => panic!("Unexpected data type {:?}", dt),
     }
+}
 
-    pub fn null_count(&self) -> usize {
-        self.null_count
-    }
+/// ----------------------------------------------------------------------------
+/// Implementations of different array types
 
-    pub fn validity_bitmap(&self) -> &Option<Bitmap> {
-        &self.validity_bitmap
-    }
+struct RawPtrBox<T> {
+    inner: *const T,
+}
 
-    pub fn get(&self, i: usize) -> &[T] {
-        self.data.get(i)
+impl<T> RawPtrBox<T> {
+    fn new(inner: *const T) -> Self {
+        Self { inner: inner }
     }
 
-    pub fn list(&self) -> &List<T> {
-        &self.data
+    fn get(&self) -> *const T {
+        self.inner
     }
 }
 
-/// Create a ListArray<T> from a List<T> without null values
-impl<T> From<List<T>> for ListArray<T>
-where
-    T: ArrowPrimitiveType,
-{
-    fn from(list: List<T>) -> Self {
-        let len = list.len();
-        ListArray {
-            len,
-            data: list,
-            validity_bitmap: None,
-            null_count: 0,
-        }
-    }
+unsafe impl<T> Send for RawPtrBox<T> {}
+unsafe impl<T> Sync for RawPtrBox<T> {}
+
+/// Array whose elements are of primitive types.
+pub struct PrimitiveArray<T: ArrowPrimitiveType> {
+    data: ArrayDataRef,
+    /// Pointer to the value array. The lifetime of this must be <= to the value buffer
+    /// stored in `data`, so it's safe to store.
+    raw_values: RawPtrBox<u8>,
+    _phantom: PhantomData<T>,
```

Comment:
Can you add a comment explaining what this is?

Reply:
Sure. Will do. It's just a marker to let the compiler know that `T` is used.

Reply:
I think it's not required anymore because of the change in `RawPtrBox`. Beforehand, it was there because the compiler requires you to use all type parameters (otherwise, rust would have some soundness problems). See https://play.rust-lang.org/?gist=32d1891decd3e0dfcdf14f92ca123e63&version=stable&mode=debug&edition=2015 for what would have happened w/o the marker.

Reply:
Yes `PhantomData` is no longer needed. Done.